### PR TITLE
Moving lisp into a separate container

### PR DESCRIPTION
**Pull request by rvs**
This one is interesting, because it presents our first case where we're moving a pillar type of a service (lisp-ztr) into a separate container. First the good news: given that our PubSub directories are now visible under /run AND given that in all EVE containers we now have /var/run -> /run everything is expected to work when it comes to PubSub. The only exception to that rule is /var/tmp/GlobalConfig PubSub endpoint, but given that /var/tmp/GlobalConfig -> /persist/... it just so happens to work as well.

Another bit of good news is that this is net-neutral for the size of the go binary. This moves ~10Mb from zedbox to a standalone lisp-ztr.

Final bit of excellent news is that from now on I can actually test lisp service in EVE as a standalone container (provided I supply /run and /persist).

Now comes a really interesting part in all of this: so far in zedrouter we used LISP specific commands to  restart lisp service. We now have an option of replacing them with generic EVE service management commands (e.g. eve restart lisp, eve stop lisp). For now, though, I just made those into no-ops:
```
        RestartCmd   = "/bin/true"
        StopCmd      = "/bin/true"
```

There's still a ton missing on the watchdog side and logmanager -- but this is good enough to be seen for the high-level feedback.

I will suggest you start your review directly with the following commit: https://github.com/lf-edge/eve/pull/607/commits/0bc8a3c2af1eaec32bb02e24b6fad521f466a2ee

### CODE REVIEW

File: pkg/lisp/go.mod [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+require (
+	github.com/google/gopacket v1.1.16
+	github.com/lf-edge/eve/pkg/pillar v0.0.0-20200203060116-d5cb7c3fb58a
+	github.com/sirupsen/logrus v1.4.2
+	golang.org/x/net v0.0.0-20191004110552-13f9640d40b9
+)
+
+replace github.com/lf-edge/eve/api/go => ../../api/go
+
+// replace github.com/lf-edge/eve/pkg/pillar => ../../../../../pkg/pillar
```
Comment (deitch) [inline pkg/lisp/go.mod]: Why does this depend on pillar? Is it some of the libs we use? And could those be moved entirely to lisp, or are they shared, e.g. pubsub?

**Comment by deitch** [comment]
I really like this, decomposing pillar, but especially some of the more complex ones like lisp. 

I have a few comments/questions:

* Dependency on pillar, per my comment on the code. What does it depend upon, and can it be pulled out?
* I am fairly certain from the work that I did that the only reason we have CGO enabled in pillar is because of the dependency on github.com/google/go-packet. Specifically, those are all in (under `pkg/pillar` before this PR: `cmd/dataplane/`, `cmd/zedrouter/`, `dataplane/*`. With this move, can we isolated `CGO` to just `pkg/lisp` and run `pkg/pillar` with `CGO_ENABLED=0`? That would simplify all of our cross-compiling dramatically.
* Can we document how this works with the change? Specifically, does the lisp package just control networking, but the data plane happens directly? Where do packets flow (which containers) and how does pillar control it? I don't get the control flows and data flows when using lisp with this change, and a short doc in `docs/`, perhaps a flow diagram or two to go with it, would be very helpful.

**Comment by rvs** [comment]
Hey @deitch I've updated the PR to account for Yetus feedback and it is ready to be merged now. Thanks for reviewing it. Now, to answer your questions:

> Dependency on pillar, per my comment on the code. What does it depend upon, and can it be pulled out?

There are legitimate dependencies on pillar when it comes lisp dataplane:
```
      pillar/watch
      pillar/pubsub
      pillar/pidfile
      pillar/agentlog
      pillar/utils
      pillar/zboot
      pillar/types
      pillar/flextimer
```

I am actually not quite sure how to manage this dependency in and automated way. Looking for suggestions here.

> I am fairly certain from the work that I did that the only reason we have CGO enabled in pillar is because of the dependency on github.com/google/go-packet. Specifically, those are all in (under pkg/pillar before this PR: cmd/dataplane/, cmd/zedrouter/, dataplane/*. With this move, can we isolated CGO to just pkg/lisp and run pkg/pillar with CGO_ENABLED=0? That would simplify all of our cross-compiling dramatically.

You have good memory man ;-) I had exactly the same hunch and was somewhat disappointed to see that we now have that dependency proliferating into one additional place:
```
pkg/pillar/cmd/zedrouter/flowstats.go
```

But I agree -- this is a much more tightly controlled dependency so perhaps @nordmark can help us come up with a way to isolate it somehow.

> Can we document how this works with the change? Specifically, does the lisp package just control networking, but the data plane happens directly? Where do packets flow (which containers) and how does pillar control it? I don't get the control flows and data flows when using lisp with this change, and a short doc in docs/, perhaps a flow diagram or two to go with it, would be very helpful.

Good point -- let me start a quick NETWORK-PLUMBING.md doc. I'll add it to the PR shortly.

**Comment by deitch** [comment]
> I am actually not quite sure how to manage this dependency in and automated way. Looking for suggestions here

The right way, IMHO, is to extract any common dependencies into a separate non-`pkg` directory, perhaps `lib`, so that `pkg/pillar` and `pkg/lisp` and `pkg/*` depend on common library modules.  Then `pkg/*` becomes running services, and `lib/*` become just dependencies.

But I think that is a bridge too far for this PR. When you are done with the first tranche of extracts from pillar, we can do a separate PR that does just that, as it will touch a lot of files, but change zero functionality.

> I had exactly the same hunch and was somewhat disappointed to see that we now have that dependency proliferating into one additional place `pkg/pillar/cmd/zedrouter/flowstats.go`

Oh, ouch. What does that single file do? Is it something that we can extract out somehow?

I looked at the dependencies, I see:

```go
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
```

`layers` has no CGO dependencies, so that leaves `pcap`, which we use only at one point, for `pcap.OpenLive()`, and then call a few functions on its returned `*pcap.Handle`: `handle.Close()`, `handle.SetBPFFilter(filter)` and `gopacket.NewPacketSource(handle, handle.LinkType())`.

This definitely narrows it down, but clearly won't eliminate it this round. Still, it may be worth the effort in a separate round. That would bring us a long way towards faster compiles, much easier cross-compile, and running pillar standalone.

> let me start a quick NETWORK-PLUMBING.md doc. I'll add it to the PR shortly

👍

**Comment by deitch** [comment]
Did you get to the doc? I don’t see it here, but it could just be because of the phone.

**Pull request merged**